bgpd/bgp_mplsvpn.c

```diff
@@ -2454,3 +2487,179 @@ void vpn_leak_postchange_all(void)
 			bgp);
 	}
 }
+
+/* When a bgp vrf instance is unconfigured, remove its routes
+ * from the VPN table and this vrf could be importing routes from other
+ * bgp vrf instnaces, unimport them.
+ * VRF X and VRF Y are exporting routes to each other.
+ * When VRF X is deleted, unimport its routes from all target vrfs,
+ * also VRF Y should unimport its routes from VRF X table.
+ * This will ensure VPN table is cleaned up appropriately.
+ */
+int bgp_vpn_leak_unimport(struct bgp *from_bgp, struct vty *vty)
+{
+	struct bgp *to_bgp;
+	const char *tmp_name;
+	char *vname;
+	struct listnode *node, *next;
+	safi_t safi = SAFI_UNICAST;
+	afi_t afi;
+	bool is_vrf_leak_bind;
+	int debug;
+
+	debug = (BGP_DEBUG(vpn, VPN_LEAK_TO_VRF) |
+		     BGP_DEBUG(vpn, VPN_LEAK_FROM_VRF));
+
+	tmp_name = from_bgp->name ? from_bgp->name : VRF_DEFAULT_NAME;
+
+	/* Default instnace can not be removed if VRF leak is enabled
+	 * across any vrf instance.
+	 */
+	if (from_bgp->inst_type == BGP_INSTANCE_TYPE_DEFAULT) {
+		if (is_bgp_vpn_leak_present()) {
+			vty_out(vty,
+			"%% VRF Leak is present, cannot remove default instance.\n");
+			return -1;
+		}
+		return 0;
+	}
+
+	if (from_bgp->inst_type != BGP_INSTANCE_TYPE_VRF)
+		return 0;
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		/* vrf leak is for IPv4 and IPv6 Unicast only */
+		if (afi != AFI_IP && afi != AFI_IP6)
+			continue;
+
+		for (ALL_LIST_ELEMENTS_RO(bm->bgp, next, to_bgp)) {
+			if (from_bgp == to_bgp)
+				continue;
+
+			/* Unimport and remove source vrf from the
+			 * other vrfs import list.
+			 */
+			struct vpn_policy *to_vpolicy;
+
+			is_vrf_leak_bind = false;
+			to_vpolicy = &(to_bgp->vpn_policy[afi]);
+			for (ALL_LIST_ELEMENTS_RO(to_vpolicy->import_vrf, node,
+						  vname)) {
+				if (strcmp(vname, tmp_name) == 0) {
+					is_vrf_leak_bind = true;
+					break;
+				}
+			}
+			/* skip this bgp instance as there is no leak to this
+			 * vrf instance.
+			 */
+			if (!is_vrf_leak_bind)
+				continue;
+
+			if (debug)
+				zlog_debug("%s: unimport routes from %s to_bgp %s afi %s import vrfs count %u",
+					   __func__, from_bgp->name_pretty,
+					   to_bgp->name_pretty, afi2str(afi),
+					   to_vpolicy->import_vrf->count);
+
+			vrf_unimport_from_vrf(to_bgp, from_bgp, afi, safi);
+
+			/* readd vrf name as unimport removes import vrf name
+			 * from the destination vrf's import list where the
+			 * `import vrf` configuration still exist.
+			 */
+			vname = XSTRDUP(MTYPE_TMP, tmp_name);
+			listnode_add(to_bgp->vpn_policy[afi].import_vrf,
```

Comment:
Chirag, I got the reason why you are doing this re-adding.

Two things.
1. I think this will still have issues. For example, in vrf_unimport_from_vrf function, after listnode_delete, if the count becomes zero, it does some cleaning by calling vpn_leak_prechange and freeing other things. In vrf_import_from_vrf function, if it is first export VRF, then the code does certain things like forming RD and setting certain flags. With the new code, none of these will happen as we simply readd this VRF again to the list. Is this okay ?
2. I feel we should avoid this "deleting and readding" business.

Reply:
@srimohans 
This solution is aligned to the current vrf route leak implementation. Anytime user does import vrf x, it is expected to clean up from vpn table followed by add.  With this change when a user does no router bgp x, it ensures to do the same task as  'no import vrf x from the vrfs where it is imported to. 
Not sure what you mean by doing certain thing to form the RD. 
In absence of this change, when user does 'import vrf x' first time the RD and RTs has to be created the source vrf and added to the destination vrfs.  
I have ensure to called unified apis across for instance when a router id change which you have reviewed it earlier.
This change is no different. 
The readd is of the vrf name to destination vrfs list is to preserve user configuration. 

@louberger any thoughts for this PR?